File: javascript/flightnerve.js
// Minimal JavaScript client for the FlightNerve flight tracking API.
//
// Docs:    https://flightnerve.com/doc/
// Get key: https://flightnerve.com/register/
//
// Node 18+ (uses the built in fetch). No dependencies. The API key goes in the
// URL path. Set it with FLIGHTNERVE_API_KEY, or pass it to new FlightNerve(...).

const BASE_URL = "https://api.flightnerve.com";

export class FlightNerveError extends Error {
  constructor(status, message) {
    super(`[${status}] ${message}`);
    this.name = "FlightNerveError";
    this.status = status;
  }
}

function clean(params) {
  const out = {};
  for (const [k, v] of Object.entries(params)) {
    if (v !== undefined && v !== null) out[k] = String(v);
  }
  return out;
}

export class FlightNerve {
  constructor(api_key, base_url = BASE_URL) {
    this.api_key = api_key || process.env.FLIGHTNERVE_API_KEY;
    if (!this.api_key) {
      throw new Error(
        "No API key. Set FLIGHTNERVE_API_KEY or pass it to new FlightNerve(key). " +
        "Get one free at https://flightnerve.com/register/"
      );
    }
    this.base_url = base_url.replace(/\/+$/, "");
  }

  async _get(endpoint, params = {}) {
    const qs = new URLSearchParams(clean(params)).toString();
    const url = `${this.base_url}/${endpoint}/${this.api_key}` + (qs ? `?${qs}` : "");
    return this._handle(await fetch(url));
  }

  async _post(endpoint, body = {}) {
    const url = `${this.base_url}/${endpoint}/${this.api_key}`;
    const res = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(clean(body)),
    });
    return this._handle(res);
  }

  async _handle(res) {
    let data;
    try {
      data = await res.json();
    } catch (e) {
      throw new FlightNerveError(res.status, "invalid JSON response");
    }
    if (!res.ok) {
      const message = (data && data.message) || "request failed";
      throw new FlightNerveError(res.status, message);
    }
    return data;
  }

  // ----- endpoints -----

  // GET /airline. Schedule, route, aircraft and live status for one flight.
  flight_status({ name, num, date, depap } = {}) {
    return this._get("airline", { name, num, date, depap });
  }

  // GET /track. Live position of an airborne flight. Empty array if not airborne.
  live_position({ name, num, date } = {}) {
    return this._get("track", { name, num, date });
  }

  // GET /schedules. Projected timetable for a flight over the coming days.
  schedules({ name, num, date, days } = {}) {
    return this._get("schedules", { name, num, date, days });
  }

  // GET /airport. Lookup by code, text search, or nearest to a coordinate.
  airport({ code, search, lat, lon, radius } = {}) {
    return this._get("airport", { code, search, lat, lon, radius });
  }

  // GET /airspace. Aircraft airborne in an area, or inbound to an airport.
  airspace({ lat, lon, radius, bbox, airport, inbound } = {}) {
    return this._get("airspace", { lat, lon, radius, bbox, airport, inbound });
  }

  // GET /route. Route detail, airline network, airport destinations, or one flight.
  route({ from, to, airline, name, num } = {}) {
    return this._get("route", { from, to, airline, name, num });
  }

  // GET /watch. Your tracked flights with scheduled, FN estimated and live arrivals.
  watchlist() {
    return this._get("watch", {});
  }

  // POST /watch. Add a flight to your watchlist.
  watch_add({ flight, date, leg } = {}) {
    return this._post("watch", { flight, date, leg });
  }

  // POST /watch with action remove. Stop tracking a flight.
  watch_remove({ flight, id } = {}) {
    return this._post("watch", { action: "remove", flight, id });
  }
}

File: python/flightnerve.py
"""
Minimal Python client for the FlightNerve flight tracking API.

Docs:    https://flightnerve.com/doc/
Get key: https://flightnerve.com/register/

The only dependency is `requests`. Your API key goes in the URL path. Set it
with the FLIGHTNERVE_API_KEY environment variable, or pass it to FlightNerve(...).
"""
import os
import requests

BASE_URL = "https://api.flightnerve.com"


class FlightNerveError(RuntimeError):
    """Raised when the API returns a non 2xx response."""
    def __init__(self, status, message):
        super().__init__("[%s] %s" % (status, message))
        self.status = status
        self.message = message


class FlightNerve:
    def __init__(self, api_key=None, base_url=BASE_URL, timeout=30):
        self.api_key = api_key or os.environ.get("FLIGHTNERVE_API_KEY")
        if not self.api_key:
            raise ValueError(
                "No API key. Set FLIGHTNERVE_API_KEY or pass api_key=... "
                "(get one free at https://flightnerve.com/register/)."
            )
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout
        self.session = requests.Session()

    def _get(self, endpoint, **params):
        params = {k: v for k, v in params.items() if v is not None}
        url = "%s/%s/%s" % (self.base_url, endpoint, self.api_key)
        return self._handle(self.session.get(url, params=params, timeout=self.timeout))

    def _post(self, endpoint, body):
        body = {k: v for k, v in body.items() if v is not None}
        url = "%s/%s/%s" % (self.base_url, endpoint, self.api_key)
        return self._handle(self.session.post(url, json=body, timeout=self.timeout))

    @staticmethod
    def _handle(resp):
        try:
            data = resp.json()
        except ValueError:
            raise FlightNerveError(resp.status_code, resp.text[:200])
        if not resp.ok:
            message = data.get("message") if isinstance(data, dict) else str(data)
            raise FlightNerveError(resp.status_code, message or "request failed")
        return data

    # ----- endpoints -----

    def flight_status(self, name, num, date=None, dep_airport=None, arr_airport=None):
        """GET /airline. Full schedule, route, aircraft and live status for one flight.
        dep_airport (depap) pins the leg leaving that airport; arr_airport (arrap) pins the leg arriving there."""
        return self._get("airline", name=name, num=num, date=date, depap=dep_airport, arrap=arr_airport)

    def live_position(self, name, num, date=None):
        """GET /track. Live position of an airborne flight. Empty list if not airborne."""
        return self._get("track", name=name, num=num, date=date)

    def schedules(self, name, num, date=None, days=None):
        """GET /schedules. Projected timetable for a flight over the coming days."""
        return self._get("schedules", name=name, num=num, date=date, days=days)

    def airport(self, code=None, search=None, lat=None, lon=None, radius=None):
        """GET /airport. Lookup by code, text search, or nearest to a coordinate."""
        return self._get("airport", code=code, search=search, lat=lat, lon=lon, radius=radius)

    def airspace(self, lat=None, lon=None, radius=None, bbox=None, airport=None, inbound=None):
        """GET /airspace. Every aircraft airborne in an area, or inbound to an airport."""
        return self._get("airspace", lat=lat, lon=lon, radius=radius,
                         bbox=bbox, airport=airport, inbound=inbound)

    def route(self, origin=None, destination=None, airline=None, name=None, num=None):
        """GET /route. Route detail, an airline network, an airport's destinations, or one flight's route."""
        return self._get("route", **{"from": origin, "to": destination,
                                     "airline": airline, "name": name, "num": num})

    def watchlist(self):
        """GET /watch. Your tracked flights, each with scheduled, FN estimated and live arrival anchors."""
        return self._get("watch")

    def watch_add(self, flight, date=None, leg=None):
        """POST /watch. Add a flight to your watchlist."""
        return self._post("watch", {"flight": flight, "date": date, "leg": leg})

    def watch_remove(self, flight=None, watch_id=None):
        """POST /watch with action remove. Stop tracking a flight."""
        return self._post("watch", {"action": "remove", "flight": flight, "id": watch_id})
